Aerogear 8470

laurafitzgerald opened the pull request:
## Motivation
https://issues.jboss.org/browse/AEROGEAR-8470

## What
Breadcrumb and Buttons

## How
Using PF4 toolbar and some css hacks

## Verification Steps
 
1. Run 'make serve'
2. The page should look as below


<img width="970" alt="Screenshot 2019-03-13 at 18 05 59" src="https://user-images.githubusercontent.com/6498727/54303342-ae2e1500-45ba-11e9-8551-79f0e4ad87a0.png">
 

3. Click on the Home breadcrumb
4. The page should navigate to the Landing Page
5. Navigation history should also work

## Checklist:

- [x] Code has been tested locally by PR requester
- [ ] Changes have been successfully verified by another team member 

## Progress

- [x] Functionality
- [x] Rebase on [Table pr](https://github.com/aerogear/mobile-security-service/pull/91) once merged
- [x] Rebase on [get request pr](https://github.com/aerogear/mobile-security-service/pull/95) once merged
- [ ] Tests

davidffrench commented on ui/src/components/AppDetailToolbar.js:
```diff
@@ -0,0 +1,64 @@
+  }
+
+  onClick = () => {
+    this.setState({
```
This will need to open up the modal if the current form state is dirty.

laurafitzgerald replied:
@davidffrench yes, this is a temporary usage until we have the the redux state implementation complete. @damienomurchu will be making some changes and @craicoverflow will be adding some changes as part of his pr. Once those are in I will be able to update this correctly.

davidffrench replied:
perfect

laurafitzgerald commented on ui/src/components/AppDetailToolbar.js:
```diff
@@ -0,0 +1,64 @@
+import './AppDetailedToolbar.css';
+
+class AppDetailToolbar extends React.Component {
+  componentWillMount () {
```
@davidffrench after the discussion yesterday I thought we weren't going to implement that everywhere yet?

davidffrench replied:
I was looking into React hooks, which we will not investigate just yet. For now we should be sticking to the high ordered component design pattern where it makes sense.

craicoverflow replied:
@laurafitzgerald @davidffrench what is the status of this, can it be reviewed?

craicoverflow commented:
I can get this to run:

```
./src/components/AppDetailedView.js
Module not found: Can't resolve './AppDetailToolbar' in '/home/endaphelan/go/src/github.com/aerogear/mobile-security-service/ui/src/components'
```

And CircleCI build [fails](https://circleci.com/gh/aerogear/mobile-security-service/1213?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link).

laurafitzgerald commented:
@craicoverflow I fixed the issue you saw. It was due to a funky rebase. Should be runnable now.

craicoverflow commented on ui/src/containers/AppDetailsToolbarContainer.js:
```diff
@@ -0,0 +1,41 @@
+import { AppDetailToolbar } from '../components/AppDetailToolbar';
+
+export class AppDetailsToolbarContainer extends React.Component {
+  onDisableButton = () => {
```
Could be named `onDisableApp` to be more descriptive. It is an event handler which could technically be used by something other that a button.

craicoverflow commented on ui/src/containers/AppDetailsToolbarContainer.js:
```diff
@@ -0,0 +1,41 @@
+    this.props.history.push('/');
+  }
+
+  onSaveButton = () => {
```
Could be named `onSaveApp` to be more descriptive. It is an event handler which could technically be used by something other that a button.

craicoverflow commented on ui/src/components/AppDetailToolbar.js:
```diff
@@ -0,0 +1,21 @@
+        <BreadcrumbItem isActive >{appName}</BreadcrumbItem>
+      </Breadcrumb>
+    </ToolbarGroup>
+    <ToolbarGroup className='toolbar-buttons'>
```
There should be a bit of space between the buttons.

craicoverflow commented on ui/src/components/AppDetailToolbar.js:
```diff
@@ -0,0 +1,21 @@
+      </Breadcrumb>
+    </ToolbarGroup>
+    <ToolbarGroup className='toolbar-buttons'>
+      <Button onClick={onDisableButton} variant="primary">Disable App</Button>
```
Actually the buttons on the Patternfly Docs aren't all caps so mabe we should be following their style?

https://patternfly-react.surge.sh/patternfly-4/components/button

laurafitzgerald replied:
Hmm, I don't think I've done anything to transform it. @davidffrench out of interest what way was the text in the buttons you used for the modal. I can't recall.

davidffrench replied:
Patternfly is capitalizing the buttons for me as well. Seems to be the default behaviour, I am not sure why this is not the case in the docs.

laurafitzgerald replied:
I found it: I added a text-transform here: https://github.com/aerogear/mobile-security-service/pull/94/files#diff-2e827c03d058d8c099d6ee8dee6693d6R6 This was to uppercase the header of the tables. This can be removed and the text can just be made uppercase in the state.

craicoverflow commented on ui/src/components/AppDetailedView.js:
```diff
@@ -16,6 +16,7 @@ class AppDetailedView extends React.Component {
     return (
       <div className="app-detailed-view">
         <Header />
+        <AppDetailsToolbarContainer />
```
I think the toolbar would look better if it was the same width as the table below.

laurafitzgerald replied:
@craicoverflow I will update the padding on the left. In the flat the buttons are more towards the corner so I've left those for now. Let me know if you want that update. Resolving this one for now unless change to the buttons margin is requested.

craicoverflow commented on ui/src/components/AppDetailToolbar.js:
```diff
@@ -0,0 +1,21 @@
+  <Toolbar className='details-toolbar'>
+    <ToolbarGroup>
+      <Breadcrumb>
+        <BreadcrumbItem onClick={onHomeClick} to='#'>Home</BreadcrumbItem>
```
The href here should be "/" not "#" as clicking this will navigate to `http://localhost:3000/#` instead of the expected `http://localhost:3000/`. The navigation history then needs 2 back-clicks to return here.

laurafitzgerald replied:
The reason I'm doing the route this way is that if I use `/` then the whole page re-renders whereas using `#` only rerenders the content so the navigatioin looks a lot cleaner. In fact the onClick is the logic which is doing the navigation. The reason for including `to='#'` is because it gives the button the correct styling. Without it it doesn't behave like a link.

craicoverflow replied:
Okay but then it messes up the navigation history so each way has a drawack. Which is better, having the whole view re-rendered and getting the nav history to work as expected, or doing a partial re-render where the user doesn't get what they expect when they press the back button?

We could not include `to='#'` and give it a custom CSS class where we style it ourselves to behave just like a link? I'm sure if we delve into the Patternfly styling we could find it and copy the class.

laurafitzgerald replied:
Yes probably the latter would be the better option as it would satisfy both. I'll have a look into that.

damienomurchu commented on ui/src/components/common/Header.css:
```diff
@@ -1,5 +1,6 @@
-.mssHeader {
+.pf-c-page__header{
```
@laurafitzgerald I wonder if using the pf internal classes is the right way to go - can you expand on your reasoning for the change?

craicoverflow replied:
Good point @damienomurchu I second that. We should not modify the internal classes, but add a second class to override some of the values.

damienomurchu commented on ui/src/components/AppDetailedToolbar.css:
```diff
@@ -0,0 +1,8 @@
+.details-toolbar{
+    margin-left: 3% 
```
Alignment & spacing seems a bit off when I view it, if we can style this to be more consistent with the spacing & alignment of the header, sub-header & table header;

![Screenshot from 2019-03-19 11-29-30](https://user-images.githubusercontent.com/12085451/54602846-acd96e00-4a3a-11e9-9067-9374a6f0bd94.png)

laurafitzgerald replied:
@damienomurchu As there is an additional section to go inbetween the breadcrumb and the table I didn't put much thought into the alignment as this may be changed by @craicoverflow's work on that.

laurafitzgerald replied:
I've made changes to align the breadcrumb with the table so resolving this one for now.

damienomurchu commented on ui/src/components/AppDetailedToolbar.css:
```diff
@@ -0,0 +1,8 @@
+.details-toolbar{
+    margin-left: 3% 
+}
+.toolbar-buttons{
```
Styling feedback:

- as already mentioned, some spacing between buttons needs to be added. 
- colour of buttons
  - should both be same colour
  - are chosen colours best to integrate with page

laurafitzgerald replied:
I've used the default styling which comes with patternfly, i.e. the primary button. We also use the blue buttons in mdc so wanted to keep that consistent. I'm aware that @davidffrench is using a red button in one of the modals so I think we do have some creative licence here.

damienomurchu commented:
@laurafitzgerald good work, basic structure seems to be in place. Some comments left in PR mainly around the styling, as well as +1'ing suggestions made by @craicoverflow

laurafitzgerald commented:
@damienomurchu @craicoverflow I've updated this pr now after rebasing, addressing feedback and adding tests. It should now be ready for a fresh code review and verification.

damienomurchu commented on ui/src/components/AppDetailsToolbar.js:
```diff
@@ -0,0 +1,39 @@
+      <Toolbar className='details-toolbar'>
+        <ToolbarGroup>
+          <ToolbarItem>
+            <Breadcrumb>
```
Breadcrumb still looks very small - can it be styled to increase the font a little?

damienomurchu commented on ui/src/components/AppDetailsToolbar.js:
```diff
@@ -0,0 +1,39 @@
+          </ToolbarItem>
+        </ToolbarGroup>
+        <ToolbarGroup className='toolbar-buttons'>
+          <Button className='toolbar-button' onClick={this.onDisableApp} variant="primary">Disable App</Button>
```
I'm just wondering should both these buttons be the same colour, or should we distinguish them in some way from each other?

laurafitzgerald replied:
There was another comment above about this also. Waiting on a reply from @davidffrench about whether we should change button colors. The standard used in mdc is blue so I think let's leave them like that for now so as not to block this pr.

damienomurchu commented on ui/src/components/AppDetailsToolbar.js:
```diff
@@ -0,0 +1,39 @@
+        </ToolbarGroup>
+        <ToolbarGroup className='toolbar-buttons'>
+          <Button className='toolbar-button' onClick={this.onDisableApp} variant="primary">Disable App</Button>
+          <Button className='toolbar-button' onClick={this.onSaveApp} variant="primary">Save</Button>
```
Re spacing on between the buttons, I see patternfly uses the following to add spacing between them, fyi
```
<Button variant="secondary">Secondary</Button>{' '}
<Button variant="tertiary">Tertiary</Button>
```

laurafitzgerald replied:
Hmm. I've tested that and it doesn't seem to work. Is it necessary to switch from the current implementation?

damienomurchu commented:
@laurafitzgerald functionally looks ok, still some issues with the styling before it's merged, imo. If this PR is blocking others maybe the styling will be resolved in another PR

davidffrench commented on ui/src/components/AppDetailedToolbar.js:
```diff
@@ -0,0 +1,39 @@
+  }
+
+  onHomeClick = () => {
+    this.props.history.push('/');
```
I would suggest switching this to `this.props.history.goBack();`

laurafitzgerald replied:
using goBack() will navigate the users path rather than back through the tree of the website so that I think we should stick with it as is.
